go/vt/vttablet/tabletmanager/restore.go

```diff
@@ -166,6 +187,105 @@ func (agent *ActionAgent) restoreDataLocked(ctx context.Context, logger logutil.
 	return nil
 }
 
+func (agent *ActionAgent) restoreToTimeFromBinlog(ctx context.Context, pos mysql.Position, restoreTime *vttime.Time) error {
+	// validate the dependent settings
+	if *binlogHost == "" || *binlogPort <= 0 || *binlogUser == "" {
+		return vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "restore_to_time flag depends on binlog server flags(binlog_host, binlog_port, binlog_user, binlog_password)")
+	}
+
+	//if restoreTime.Seconds > int64(time.Now().Second()) {
+	//	println(restoreTime.Seconds)
+	//	println(time.Now().Second())
+	//	log.Warning("Restore time request is a future date, so skipping it")
+	//	return nil
+	//}
+	println("restoring from below time")
+	println(restoreTime.Seconds)
+	gtid := agent.getGTIDFromTimestamp(ctx, pos, restoreTime.Seconds)
+	if gtid == "" {
+		return vterrors.New(vtrpcpb.Code_FAILED_PRECONDITION, "unable to fetch the GTID for the specified restore_to_time")
+	}
+
+	println(fmt.Sprintf("going to restore upto the gtid - %s", gtid))
+	err := agent.catchupToGTID(ctx, gtid)
+	if err != nil {
+		return vterrors.Wrapf(err, "unable to replicate upto specified gtid : %s", gtid)
+	}
+
+	return nil
+}
+
+// getGTIDFromTimestamp gets the next GTID of the event happened on the timestamp (resore_to_time)
+func (agent *ActionAgent) getGTIDFromTimestamp(ctx context.Context, pos mysql.Position, restoreTime int64) string {
+	connParams := &mysql.ConnParams{
+		Host:  *binlogHost,
+		Port:  *binlogPort,
+		Uname: *binlogUser,
+		Pass:  *binlogPwd,
+	}
+	dbCfgs := &dbconfigs.DBConfigs{
+		Host: connParams.Host,
+		Port: connParams.Port,
+	}
+	dbCfgs.SetDbParams(*connParams, *connParams)
+	vsClient := vreplication.NewReplicaConnector(connParams)
+
+	filter := &binlogdatapb.Filter{
+		Rules: []*binlogdatapb.Rule{{
+			Match: "/.*",
+		}},
+	}
+
+	// Todo: we need to safely return from vstream if it takes more time
+	found := make(chan string)
+	go func() {
+		err := vsClient.VStream(ctx, mysql.EncodePosition(pos), filter, func(events []*binlogdatapb.VEvent) error {
+			for _, event := range events {
+				if event.Gtid != "" && event.Timestamp > restoreTime {
+					found <- event.Gtid
+					break
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			found <- ""
+		}
+	}()
+	timeout := time.Now().Add(60 * time.Second)
+	for time.Now().Before(timeout) {
+		select {
+		case <-found:
+			return <-found
+		default:
+			time.Sleep(300 * time.Millisecond)
+		}
+	}
+	return ""
+}
+
+// replicateUptoGTID replicates upto specified gtid from binlog server
+func (agent *ActionAgent) catchupToGTID(ctx context.Context, gtid string) error {
+	gtid = strings.Replace(gtid, "MySQL56/", "", 1)
+
+	gtidNew := strings.Split(gtid, ":")[0] + ":" + strings.Split(strings.Split(gtid, ":")[1], "-")[1]
```

Comment:
take a look at https://github.com/vitessio/vitess/blob/master/go/mysql/mysql56_gtid_set.go. maybe you don't need to do all this manual processing of the GTID string.

Reply:
yes, I was planning to do it, this is a quick hack to get started and make it working.